diff --git a/common/common.cpp b/common/common.cpp
index a10685f1..9ecbf62a 100644
--- a/common/common.cpp
+++ b/common/common.cpp
@@ -1652,31 +1652,26 @@ bool gpt_params_find_arg(int argc, char ** argv, const std::string & arg, gpt_pa
     }
     if (arg == "--allowlist-unicode-rule") {
         CHECK_ARG
-        params.allow_rules.push_back(argparse_allowlist_unicode_rule(argv[i]));
-        return true;
-    }
-    if (arg == "--allowlist-no-defer-common") {
-        params.allow_df_common = false;
-        return true;
-    }
-    if (arg == "--allowlist-each-piece") {
-        CHECK_ARG
-        params.allow_each_pieces.push_back(argv[i]);
+        if (params.allow_ruless.size() == 0) {
+            params.allow_ruless.push_back({});
+        }
+        params.allow_ruless.back().push_back(argparse_allowlist_unicode_rule(argv[i]));
         return true;
     }
-    if (arg == "--allowlist-piece") {
+    if (arg == "--allowlist-pieces") {
         CHECK_ARG
         params.allow_pieces.push_back(argv[i]);
         return true;
     }
-    if (arg == "--allowlist-binning-keyword") {
+    if (arg == "--allowlist-keyword") {
         CHECK_ARG
-        params.allow_bin_kw = argv[i];
+        params.allow_kws.push_back(argv[i]);
+        params.allow_ruless.push_back({});
         return true;
     }
-    if (arg == "--allowlist-binning-threshold") {
+    if (arg == "--allowlist-keyword-delay") {
         CHECK_ARG
-        params.allow_bin_thresh = std::stoul(argv[i]);
+        params.allow_kw_delay = std::stoul(argv[i]);
         return true;
     }
     if (arg == "-ld" || arg == "--logdir") {
@@ -2428,20 +2423,11 @@ void gpt_params_print_usage(int /*argc*/, char ** argv, const gpt_params & param
     options.push_back({ "*",           "       --banned-n",             "number of tokens banned in the phrase during rewind. -1 means all tokens: (default: %d)",params.banned_n });
     options.push_back({ "*",           "       --allowlist-unicode-rule",
                                                                         "rule for allowlisting unicode script and/or codepoints. disabled without any rule. format: `LOWER..UPPER,SCRIPT:BIAS`\n"
-                                                                        "if unspecified: LOWER = 0, UPPER = -1(=max), SCRIPT=\"\", BIAS = 0. at least one of LOWER, UPPER, or SCRIPT is required\n"
-                                                                        "tokens containing characters without applicable rule is blacklisted (BIAS=-999)\n"
-                                                                        "`--allowlist-unicode-rule ascii` is valid as special case\n"
-                                                                        "e.g. `--allowlist-unicode-rule ascii:42` == `--allowlist-unicode-rule 0..127,common:42 --allowlist-unicode-rule 0..127,latin:42`" });
-    options.push_back({ "*",           "       --allowlist-no-defer-common",
-                                                                        "common characters not in any rule defer to other characters in the token instead of blacklisting (default: %s)", params.allow_df_common ? "true" : "false" });
-    options.push_back({ "*",           "       --allowlist-each-piece", "allowlist each token in argument. inherits max BIAS in --allowlist-unicode-rule. overrides --allowlist-unicode-rule" });
-    options.push_back({ "*",           "       --allowlist-piece",      "allowlist if and only if argument tokenizes to a single token\n"
-                                                                        "inherits max BIAS in --allowlist-unicode-rule. overrides --allowlist-unicode-rule and --allowlist-each-piece" });
-    options.push_back({ "*",           "       --allowlist-binning-keyword",
-                                                                        "keyword to trigger allowlist binning if matched during generation (default: \"%s\", \"\" = disabled)\n"
-                                                                        "binning: BIAS>=0 -> 0, BIAS<0 -> -999", params.allow_bin_kw.c_str() });
-    options.push_back({ "*",           "       --allowlist-binning-threshold",
-                                                                        "# keyword match > threshold triggers binning (default: %zu)", params.allow_bin_thresh });
+                                                                        "if unspecified: LOWER = 0, UPPER = -1(=max), SCRIPT=\"\", BIAS = 0. at least one of LOWER, UPPER, or SCRIPT is required\n" });
+    options.push_back({ "*",           "       --allowlist-pieces",     "allowlist each token in argument. inherits max BIAS in --allowlist-unicode-rule. overrides --allowlist-unicode-rule" });
+    options.push_back({ "*",           "       --allowlist-keyword",    "keyword to expire earlier allowlist rules if matched during generation. does not affect later rules" });
+    options.push_back({ "*",           "       --allowlist-keyword-delay",
+                                                                        "# tokens to delay matching for the first keyword (default: %zu)", params.allow_kw_delay });
     options.push_back({ "*",           "       -l TOKEN_ID(+/-)BIAS",   "modifies the likelihood of token appearing in the completion,\n"
                                                                         "i.e. `--logit-bias 15043+1` to increase likelihood of token ' Hello',\n"
                                                                         "or `--logit-bias 15043-1` to decrease likelihood of token ' Hello'" });
diff --git a/common/common.h b/common/common.h
index ff596d6e..156920e9 100644
--- a/common/common.h
+++ b/common/common.h
@@ -286,12 +286,15 @@ struct gpt_params {
     size_t n_buffer 				 =  0; // number of token buffers for string ban
     bool can_ban_phrases             = true;  // whether to ban strings
 
-    std::vector<std::tuple<uint32_t, uint32_t, std::string, float>> allow_rules;    // rules for unicode allowlist
-    std::vector<std::string> allow_each_pieces;     // piece -> tokens -> allowlist
-    std::vector<std::string> allow_pieces;          // allowlist iff piece is single token
-    bool allow_df_common = true;    // common characters without rule defer to other characters in a token
-    std::string allow_bin_kw;       // binning keyword
-    size_t allow_bin_thresh = 0;    // keyword match threshold for binning
+    std::vector<std::vector<std::tuple<
+        uint32_t        // lower codepoint
+        ,uint32_t       // upper codepoint
+        ,std::string    // unicode script name
+        ,float          // bias
+    >>> allow_ruless;
+    std::vector<std::string> allow_pieces;  // each token to allowlist
+    std::vector<std::string> allow_kws;     // keywords
+    size_t allow_kw_delay;  // minimum n_decoded before first keyword is active
 
     std::vector<llama_model_kv_override> kv_overrides;
     std::vector<llama_model_tensor_buft_override> tensor_buft_overrides;
diff --git a/examples/server/server-context.cpp b/examples/server/server-context.cpp
index b9fb0312..653f5b23 100644
--- a/examples/server/server-context.cpp
+++ b/examples/server/server-context.cpp
@@ -389,13 +389,11 @@ void server_slot::reset() {
     ban_regex.clear();
     ban_regex_ci.clear();
 
-    allow_rules.clear();
-    allow_df_common = false;
-    allow_each_pieces.clear();
+    allow_ruless.clear();
     allow_pieces.clear();
-    allow_bin_kw = "";
-    allow_bin_thresh = 0;
-    allow_bin_kw_counter = 0;
+    allow_kws.clear();
+    allow_kw_delay = 0;
+    allow_idx = 0;
 
     // Reset speculative decoding stats
     n_draft_total = 0;
@@ -1356,36 +1354,9 @@ bool server_context::launch_slot_with_task(server_slot& slot, server_task& task)
 
     do  // populate allowlist biases
     {
-        slot.allow_rules = params_base.allow_rules;
-        slot.allow_df_common = params_base.allow_df_common;
-        const auto& allowlist_unicode_rule_array = data.find("allowlist_unicode_rule_array");
-        if ((allowlist_unicode_rule_array != data.end()) && allowlist_unicode_rule_array->is_array()) {
-            slot.allow_rules.clear();
-            for (size_t j = 0; j < allowlist_unicode_rule_array->size(); ++j) {
-                if (!allowlist_unicode_rule_array->at(j).is_string()) {
-                    continue;
-                }
-                slot.allow_rules.push_back(argparse_allowlist_unicode_rule(allowlist_unicode_rule_array->at(j).get<std::string>()));
-            }
-            slot.allow_df_common = data.find("allowlist_no_defer_common") == data.end();
-        }
-
-        if (slot.allow_rules.empty()) {
-            slot.allow_biases.clear();
-            slot.allow_bin_biases.clear();
-            break;
-        }
-
-        slot.allow_each_pieces = params_base.allow_each_pieces;
-        const auto& allowlist_each_piece_array = data.find("allowlist_each_piece_array");
-        if (allowlist_each_piece_array != data.end() && allowlist_each_piece_array->is_array()) {
-            slot.allow_each_pieces.clear();
-            for (const auto& piece: *allowlist_each_piece_array) {
-                if (piece.is_string()) {
-                    slot.allow_each_pieces.push_back(piece.get<std::string>());
-                }
-            }
-        }
+        // TODO: JSON parsing for rules and keywords
+        slot.allow_ruless = params_base.allow_ruless;
+        slot.allow_kws = params_base.allow_kws;
 
         slot.allow_pieces = params_base.allow_pieces;
         const auto& allowlist_piece_array = data.find("allowlist_piece_array");
@@ -1398,15 +1369,38 @@ bool server_context::launch_slot_with_task(server_slot& slot, server_task& task)
             }
         }
 
-        slot.allow_bin_kw = json_value(data, "allowlist_binning_keyword", params_base.allow_bin_kw);
-        slot.allow_bin_thresh = json_value(data, "allowlist_binning_threshold", params_base.allow_bin_thresh);
+        slot.allow_kw_delay = json_value(data, "allowlist_keyword_delay", params_base.allow_kw_delay);
         // end of allowlist criteria update
 
         const int32_t n_vocab = populate_vocab_pieces();
 
-        if ((slot.allow_rules != slot.allow_rules_prev) && (slot.allow_df_common ^ slot.allow_df_common_prev)) {
-            LLAMA_LOG_DEBUG("%s: applying new allowlist rules\n", __func__);
-            slot.allow_biases.resize(n_vocab);
+        std::unordered_set<llama_token> allow_settoken;
+        for (const auto& piece: slot.allow_pieces) {
+            for (const auto token: common_tokenize(model, piece, false, true)) {
+                allow_settoken.insert(token);
+            }
+        }
+
+        auto n_rules = slot.allow_ruless.size();
+        if (n_rules > slot.allow_kws.size() + 1) {
+            // one more rules than keyword, last rules do not expire
+            n_rules = slot.allow_kws.size() + 1;
+            slot.allow_ruless.resize(n_rules);
+        } else if (n_rules < slot.allow_kws.size()) {
+            // every rules expire
+            slot.allow_kws.resize(n_rules);
+        }
+        slot.allow_biasess.resize(n_rules);
+
+        for (size_t i = 0; i < n_rules; ++i) {
+            const auto& rules = slot.allow_ruless[i];
+            if ((i < slot.allow_ruless_prev.size()) && (rules == slot.allow_ruless_prev[i])) {
+                continue;
+            }
+            LLAMA_LOG_DEBUG("%s: allowlist %zu is new\n", __func__, i);
+
+            auto& biases = slot.allow_biasess[i];
+            biases.resize(n_vocab);
 
             std::vector<uint32_t> cpts;
             std::vector<std::string> scripts;
@@ -1419,7 +1413,7 @@ bool server_context::launch_slot_with_task(server_slot& slot, server_task& task)
                     bool in_rule = false;
 
                     // at least one rule
-                    for (const auto& rule: slot.allow_rules) {
+                    for (const auto& rule: rules) {
                         const bool in_range = (std::get<0>(rule) <= cpts[j]) && (cpts[j] <= std::get<1>(rule));
                         in_rule = in_range && ((std::get<2>(rule) == "*") || std::get<2>(rule) == scripts[j]);
                         if (in_rule) {
@@ -1429,8 +1423,8 @@ bool server_context::launch_slot_with_task(server_slot& slot, server_task& task)
                         }
                     }
                     if (!in_rule) {
-                        if (slot.allow_df_common && (scripts[j] == "common")) {
-                            // for common codepoints (e.g. whitespace), defer to other codepoints in the token
+                        if ((scripts[j] == "common") || (scripts[j] == "inherited")) {
+                            // for common or inherited codepoints (e.g. whitespace), defer to other codepoints in the token
                             continue;
                         }
 
@@ -1439,40 +1433,19 @@ bool server_context::launch_slot_with_task(server_slot& slot, server_task& task)
                         break;
                     }
                 }
-                slot.allow_biases[id] = bias;
+                biases[id] = bias;
             }
-        }
 
-        float max_bias = -INFINITY;
-        for (const auto& rule: slot.allow_rules) {
-            max_bias = std::max(max_bias, std::get<3>(rule));
-        }
-        for (const auto& piece: slot.allow_each_pieces) {
-            for (const auto token: common_tokenize(model, piece, false, true)) {
-                slot.allow_biases[token] = max_bias;
-            }
-        }
-        for (const auto& piece: slot.allow_pieces) {
-            const auto tokens = common_tokenize(model, piece, false, true);
-            if (tokens.size() == 1) {
-                slot.allow_biases[tokens[0]] = max_bias;
-            } else {
-                LLAMA_LOG_WARN("%s: skipping `%s`, .size = %zu\n", __func__, piece.c_str(), tokens.size());
+            float max_bias = -INFINITY;
+            for (const auto& rule: rules) {
+                max_bias = std::max(max_bias, std::get<3>(rule));
             }
-        }
-
-        if (!slot.allow_bin_kw.empty()) {
-            LLAMA_LOG_INFO("%s: preparing binned allowlist\n", __func__);
-            slot.allow_bin_biases.resize(n_vocab);
-            for (int32_t id = 0; id < n_vocab; ++id) {
-                if (slot.allow_biases[id] < 0) {
-                    slot.allow_bin_biases[id] = -999;
-                }
+            for (const auto token: allow_settoken) {
+                biases[token] = max_bias;
             }
         }
     } while (false);
-    slot.allow_rules_prev = slot.allow_rules;
-    slot.allow_df_common_prev = slot.allow_df_common;
+    slot.allow_ruless_prev = slot.allow_ruless;
 
     if (llama_model_has_recurrent(llama_get_model(slot.ctx))) {
         params_base.can_ban_phrases = false;
@@ -2070,9 +2043,11 @@ void server_context::send_embedding(const server_slot& slot, const llama_batch&
 void server_context::apply_server_biases(server_slot& slot) {
     auto& server_biases = slot.ctx_sampling->server_biases;
 
-    server_biases = slot.allow_bin_kw_counter > slot.allow_bin_thresh
-        ? &slot.allow_bin_biases
-        : &slot.allow_biases;
+    if (slot.allow_idx < slot.allow_biasess.size()) {
+        server_biases = &slot.allow_biasess[slot.allow_idx];
+    } else {
+        server_biases = nullptr;
+    }
 }
 
 void server_context::request_completion(int id_task, int id_multi, json data, bool infill, bool embedding, server_tokens&& inputs) {
@@ -3555,7 +3530,6 @@ void server_context::speculative_decoding_accept() {
 
         size_t n_draft = slot.drafted.size();
 
-        update_allowlist_state(slot);
         apply_server_biases(slot);
 
         // the accepted tokens from the speculation
@@ -3635,6 +3609,8 @@ void server_context::speculative_decoding_accept() {
             }
 
             common_sampler_review(slot.ctx_sampling, slot.token_buffer.size(), slot.rewind_status);
+
+            update_allowlist_state(slot);
         }
         SLT_DBG(slot, "accepted %d/%d draft tokens, new n_tokens = %d\n", (int)ids.size() - 1, (int)slot.drafted.size(), slot.n_past);
         LOG_VERBOSE("speculative decoding result", {
@@ -3903,21 +3879,20 @@ void server_context::buffer_and_check_string_ban(server_slot & slot, completion_
 }
 
 void server_context::update_allowlist_state(server_slot& slot) {
-    const auto& kw = slot.allow_bin_kw;
-    if (kw.empty()) {
-        // disabled
-        return;
-    }
-
-    auto& kw_counter = slot.allow_bin_kw_counter;
-    if (kw_counter > slot.allow_bin_thresh) {
+    const auto& kws = slot.allow_kws;
+    auto& idx = slot.allow_idx;
+    if ((slot.allow_kw_delay > slot.n_decoded) || (idx >= kws.size())) {
         return;
     }
 
     // search for keyword
-    auto pos = slot.generated_text.find(kw, slot.last_gentxt_size);
+    auto kw = kws[idx];
+    auto pos = slot.generated_text.find(kw, std::max(0, slot.last_gentxt_size - (int32_t)kw.length() + 1));
     while (pos != std::string::npos) {
-        ++kw_counter;
+        if (++idx >= kws.size()) {
+            break;
+        }
+        kw = kws[idx];
         pos = slot.generated_text.find(kw, pos + 1);
     }
 }
@@ -4042,9 +4017,6 @@ void server_context::process_batch_tokens(int32_t & n_batch) {
                 }
             }
 
-            update_allowlist_state(slot);
-            apply_server_biases(slot);
-
             completion_token_output result;
             const int tok_idx = slot.i_batch - i;
 
@@ -4057,6 +4029,8 @@ void server_context::process_batch_tokens(int32_t & n_batch) {
                 }
             }
 
+            apply_server_biases(slot);
+
             const llama_token id = common_sampler_sample(slot.ctx_sampling, ctx, tok_idx);
 
             common_sampler_accept(slot.ctx_sampling, ctx, id, true);
@@ -4100,6 +4074,8 @@ void server_context::process_batch_tokens(int32_t & n_batch) {
 
             common_sampler_review(slot.ctx_sampling, slot.token_buffer.size(), slot.rewind_status);
 
+            update_allowlist_state(slot);
+
             slot.i_batch = -1;
         }
         if (mtp_warmup_needed && !batch_mtp_hidden_state.empty()) {
diff --git a/examples/server/server-context.h b/examples/server/server-context.h
index 18cd0bd1..50e05131 100644
--- a/examples/server/server-context.h
+++ b/examples/server/server-context.h
@@ -64,7 +64,7 @@ struct server_slot {
     server_tokens prompt_tokens;
     server_tokens cache_tokens;
 
-    size_t last_gentxt_size = 0;
+    int32_t last_gentxt_size = 0;
     std::string generated_text;
 
     // idx of draft tokens in the main batch
@@ -104,17 +104,13 @@ struct server_slot {
 	std::map<int32_t, std::set<llama_token>> positional_bans;
 
     // allowlist
-    std::vector<std::tuple<uint32_t, uint32_t, std::string, float>> allow_rules;
-    std::vector<std::tuple<uint32_t, uint32_t, std::string, float>> allow_rules_prev;
-    bool allow_df_common = false;
-    bool allow_df_common_prev = false;
-    std::vector<std::string> allow_each_pieces;
+    std::vector<std::vector<std::tuple<uint32_t, uint32_t, std::string, float>>> allow_ruless_prev;
+    std::vector<std::vector<std::tuple<uint32_t, uint32_t, std::string, float>>> allow_ruless;
     std::vector<std::string> allow_pieces;
-    std::string allow_bin_kw;
-    size_t allow_bin_thresh = 0;
-    size_t allow_bin_kw_counter = 0;
-    std::vector<float> allow_biases;
-    std::vector<float> allow_bin_biases;
+    std::vector<std::string> allow_kws;
+    size_t allow_kw_delay = 0;
+    std::vector<std::vector<float>> allow_biasess;
+    size_t allow_idx = 0;
 
     server_prompt server_cached_prompt;
 
